frontera/contrib/backends/cassandra/components.py

```diff
@@ -0,0 +1,359 @@
+                sleep(5)
+                tries -= 1
+                if tries > 0:
+                    self.logger.info("Tries left %i", tries)
+                    continue
+                else:
+                    raise exc
+    return func_wrapper
+
+
+class Metadata(BaseMetadata):
+    def __init__(self, session_cls, model_cls, cache_size, crawl_id='default'):
+        self.session = session_cls
+        self.model = model_cls
+        self.table = 'MetadataModel'
+        self.cache = LRUCache(cache_size)
+        self.logger = logging.getLogger("frontera.contrib.backends.cassandra.components.Metadata")
+        self.crawl_id = crawl_id
+
+    def frontier_stop(self):
+        pass
+
+    @retry_and_rollback
+    def add_seeds(self, seeds):
+        cql_items = []
+        for seed in seeds:
+            o = self._create_page(seed)
+            self.cache[o.fingerprint] = o
+
+            query = self.session.prepare(
+                "INSERT INTO metadata (crawl, fingerprint, url, created_at, meta, headers, cookies, method, depth) "
+                "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)")
+            meta = Meta(domain=seed.meta['domain'], fingerprint=seed.meta['fingerprint'],
+                        origin_is_frontier=seed.meta['origin_is_frontier'],
+                        scrapy_callback=seed.meta['scrapy_callback'], scrapy_errback=seed.meta['scrapy_errback'],
+                        scrapy_meta=seed.meta['scrapy_meta'])
+            cql_i = (self.crawl_id, seed.meta['fingerprint'], seed.url, datetime.utcnow(), meta,
+                     seed.headers, seed.cookies, seed.method, o.depth)
+            cql_items.append(cql_i)
+        if len(seeds) > 0:
+            execute_concurrent_with_args(self.session, query, cql_items, concurrency=400)
+        self.cass_count({"seed_urls": len(seeds)})
+
+    @retry_and_rollback
+    def request_error(self, page, error):
+        m = self._create_page(page)
+        m.error = error
+        self.cache[m.fingerprint] = m
+        query_page = self.session.prepare(
+            "UPDATE metadata SET error = ? WHERE fingerprint = ?")
+        self.session.execute(query_page, (error, page.meta['fingerprint']))
+        self.cass_count({"error": 1})
+
+    @retry_and_rollback
+    def page_crawled(self, response, links):
+        query_page = self.session.prepare(
+            "UPDATE metadata SET fetched_at = ?, headers = ?, method = ?, cookies = ?, status_code = ? "
+            "WHERE crawl= ? AND fingerprint = ?")
+        self.session.execute_async(query_page, (datetime.utcnow(), response.request.headers, response.request.method,
+                                                response.request.cookies, response.status_code, self.crawl_id,
+                                                response.meta['fingerprint']))
+
+        query = self.session.prepare(
+            "INSERT INTO metadata (crawl, fingerprint, created_at, method, url, depth) VALUES (?, ?, ?, ?, ?, ?)")
```

Comment:
What if link already exists in the table? Will it throw Exception or silently replace the row?